components/net/lwip_telnet/telnet.c

```diff
@@ -0,0 +1,432 @@
+                else if (*data != '\r') /* ignore '\r' */
+                {
+                    rt_sem_take(telnet->rx_ringbuffer_lock, RT_WAITING_FOREVER);
+                    /* put buffer to ringbuffer */
+                    rt_ringbuffer_putchar(&(telnet->rx_ringbuffer), *data);
+                    rt_sem_release(telnet->rx_ringbuffer_lock);
+                }
+                break;
+        }
+
+        data ++;
+    }
+
+    rt_sem_take(telnet->rx_ringbuffer_lock, RT_WAITING_FOREVER);
+    /* get total size */
+    rx_length = rt_ringbuffer_data_len(&telnet->rx_ringbuffer);
+    rt_sem_release(telnet->rx_ringbuffer_lock);
+
+    /* indicate there are reception data */
+    if ((rx_length > 0) && (telnet->device.rx_indicate != RT_NULL))
+        telnet->device.rx_indicate(&telnet->device,
+                                   rx_length);
+
+    return;
+}
+
+static rt_int32_t rtu_telnet_srv_tcp_init(const rt_int16_t local_port ,
+                                   rt_int32_t nb_connection)
+{
+    rt_uint8_t data_buffer[MAX_CMD_LENS] = {0};
+    struct sockaddr_in  server_addr;
+    socklen_t addr_len;
+    struct sockaddr_in conn_addr;
+
+    int rc           = 0 ;
+    int optval       = 1;
+    int sock_fd      = -1  ;
+    int sock_conn    = -1  ;
+
+    fd_set readset;
+    struct timeval timeout;
+
+    timeout.tv_sec = 0;         // second.
+    timeout.tv_usec = 50 * 1000; //wait micro second.
+
+    sock_fd = socket(AF_INET, SOCK_STREAM, 0);
+
+    if (sock_fd == -1 )
+    {
+        return -RT_ERROR;
+    }
+
+    setsockopt(sock_fd, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));
+    setsockopt(sock_fd, SOL_SOCKET , SO_KEEPALIVE, &optval, sizeof(optval));
+
+    rt_memset(&server_addr, 0, sizeof(server_addr));
+
+    server_addr.sin_family = AF_INET;
+    server_addr.sin_port = htons(local_port);
+    server_addr.sin_addr.s_addr = htonl(INADDR_ANY);
+    rt_memset(&(server_addr.sin_zero), 0, sizeof(server_addr.sin_zero));
+
+    if (bind(sock_fd, (struct sockaddr *)&server_addr,
+             sizeof(struct sockaddr)) == -1)
+    {
+        closesocket(sock_fd);
+
+        return -RT_ERROR;
+    }
+
+    if ((listen(sock_fd, nb_connection)) == -1)
+    {
+        closesocket(sock_fd);
+
+        return -RT_ERROR;
+    }
+
+    addr_len = sizeof(struct sockaddr_in);
+
+    while(1)
+    {
+        sock_conn = accept(sock_fd, (struct sockaddr *)&conn_addr, &addr_len);
+
+        if(sock_conn > 0)
+        {
+            telnet->default_console = rt_console_get_device();
+          
+            rt_console_set_device("telnet");
+            finsh_set_device("telnet");
+            
+            /* set init state */
+            telnet->state = STATE_NORMAL;
+
+            /* disable echo mode */
+            finsh_set_echo(1);
+
+            telnet->echo_mode = finsh_get_echo();
+
+            while (1)
+            {
+                telnet_process_tx(telnet , sock_conn);
```

Comment:
/*make select function do not block*/
    timeout.tv_sec = 0;         // second.
    timeout.tv_usec = 0;        //wait micro second.